Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,309 @@
+/-
+Copyright (c) 2025 Sihan Su. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca, Sihan Su, Wan Lin, Xiaoyang Su
+-/
+import Mathlib.Algebra.MvPolynomial.Monad
+import Mathlib.Data.List.Indexes
+import Mathlib.RingTheory.IntegralClosure.IsIntegralClosure.Basic
+/-!
+# Noether Normalizaton
+This file contains a proof by Nagata of the Noether normalization lemma.
+
+## Main Results
+Let `A` be a finitely generated algebra over a field `k`.
+Then there exists a natural number `r` and an injective homomorphism
+from `k[X_1, X_2, ..., X_r]` to `A` such that `A` is integral over `k[X_1, X_2, ..., X_r]`.
+
+## Reference
+* <https://stacks.math.columbia.edu/tag/00OW>
+
+-/
+open Polynomial MvPolynomial Ideal BigOperators Nat RingHom List
+
+variable {k : Type*} [Field k] {n : ℕ} (f : MvPolynomial (Fin (n + 1)) k)
+variable (v w : Fin (n + 1) →₀ ℕ)
+
+namespace NoetherNormalization
+
+section equivT
+/-!
+Suppose `f` is a nonzero polynomial of `n+1` variables : `X_0,...,X_n`.
+First we construct an algebra equivalence `T` from `k[X_0,...,X_n]` to `k[X_0,...,X_n]`, such that
+  `f` is mapped to a polynomial of `X_0` with invertible leading coefficient.
+More precisely, `T` maps `X_i` to `X_i + X_0 ^ r_i` when `i ≠ 0`, and `X_0` to `X_0`.
+Here we choose `r_i` to be `up ^ i` where `up` is big enought, so that `T` maps
+  different monomials of `f` to mvpolynomials with different degrees of `X_0`. See `Nat.ofDigits`.
```

Comment:
```suggestion
Here we choose `r_i` to be `up ^ i` where `up` is big enought, so that `T` maps
  the different monomials of `f` to polynomials with different degrees in `X_0`. See `Nat.ofDigits`.
```
The name `Nat.ofDigits` now changed, right?